lib/segment/src/utils/maybe_arc.rs

```diff
@@ -0,0 +1,111 @@
+use std::ops::Deref;
+use std::sync::Arc;
+
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+// Structure that acts as `T` most of the time but allows to interchange being wrapped within an `Arc` or not.
+// This is helpful, when a variable can become memory-intensive but must remain the ability to get cloned.
+#[derive(Debug, Deserialize, Serialize, JsonSchema, Clone, PartialEq, Eq)]
+#[serde(untagged)] // Make this type transparent when de/serializing and always deserialize as `NoArc`, since it's the first enum kind that matches.
+pub enum MaybeArc<T> {
+    NoArc(T),
+    Arc(Arc<T>),
+}
+
+impl<T> MaybeArc<T> {
```

Comment:
A nice to have would be to choose arc/no_arc automatically based on the `size_of_value()` of the inner value as a constructor function of this type.

Based on the provided chart, and considering `ExtendedPointId` is 24 bytes, the threshold could be around 1536 bytes (24 * 64)

Reply:
`size_of_value` doesn't work with HashSet though. When using `MaybeArc` wrapped around HashMaps or Vec, `size_of_value` would only return the stack size of the type, which is always constant regardless of the amount of items.
Therefore we can only apply this heuristic at each usage of `MaybeArc` instead inside a constructor, sadly.